instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py

```diff
@@ -72,20 +72,33 @@ def __init__(self, tracer, engine):
         listen(engine, "after_cursor_execute", self._after_cur_exec)
         listen(engine, "handle_error", self._handle_error)
 
+    def _operation_name(self, db_name="", statement=""):
+        parts = []
+        if isinstance(statement, str):
+            parts.append(statement.split(" ")[0].split("\t")[0])
```

Comment:
This doesn't seem to follow the [semantic convention recommendations](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md#semantic-conventions-for-database-client-calls):

> SQL spans SHOULD be named the
following way, unless the statement is known to be of low cardinality:
`<db.operation> <db.name>.<db.sql.table>`, provided that `db.operation` and `db.sql.table` are available.
If `db.sql.table` is not available due to its semantics, the span SHOULD be named `<db.operation> <db.name>`.
It is not recommended to attempt any client-side parsing of `db.statement` just to get these properties,
they should only be used if the library being instrumented already provides them.
When it's otherwise impossible to get any meaningful span name, `db.name` or the tech-specific database name MAY be used.

I would have thought SQLAlchemy could provide a way to get `db.operation`. I'm also not too confident this parsing would always work to get the operation, e.g. if the whole query is wrapped in parens

Reply:
We had the same discussion on another PR. https://github.com/open-telemetry/opentelemetry-python-contrib/pull/246#discussion_r542562273

Copying my reply here. 

> I agree that we shouldn't parse it either but I don't think this really counts as "SQL parsing". We are making a bare-bones, extremely cheap, best effort attempt at guessing the SQL operation name. I think as long as it does not incur a perf penalty, decrease reliabilit, affect the DB connection/cursor/statement in any way or end up talking to the database, we should be good. At least that's how I understand the intention behind the spec recommendation but may be there is more to it that I might have overlooked?

Reply:
I agree that this is kind of going against the semantic conventions as well. I do not believe we should try to parse the statement to obtain <db.operation>. Ideally, this should have been addressed in [this pr](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/202) but might have been missed. We should be seeing if <db.operation> is provided by the sqlalchemy library. If not, we should fallback on <db.name>.

Reply:
I still don't think this goes against the spec. This definitely complied with the semantic conventions. Only point of contention is that the spec prescribes to not do any SQL parsing in order figure out the SQL command, table name etc. I don't think we are really doing that. If we ran an SQL parser that parsed the raw statement and tried to understand SQL grammer to figure out the SQL command and table name, I'd say that'd be against the spec. Here we are merely truncating the raw statement to the first word. I don't think that really counts as SQL parsing. If SQLAlchemy provided us with the query as a list with broken down components, it would be fine to just take the first item from the list and use it as the statement. I think what we are doing here is a lot closer to that than any kind of SQL parsing. 

Also, if we try to understand the sentiment behind the recommendation, I think the reasons to not parse are one or all of the following:

- SQL parsing can be slow, perhaps too slow for instrumentation to do.
- Depending on the library/ORM being instrumented, parsing a statement could be a stateful operation. Parsing a query once could result in a 2nd parse attempt by the library itself failing.
- Depending on the library/ORM, parsing could result in network op/re-connection or additional SQL query to retrieve metadata.
- We'd have to support a huge number of dialects and have fallbacks.
- In general, it can be error prone and make the instrumentation unreliable.

I don't think what we are doing anything here that touches on any of the issues. 

Moreover, most python SQL libraries/ORMs do not provide us with broken down information so not even truncating the raw statement to the first word gives us spans where operation name is always the same as the database name. This is very limiting for APM backends as they cannot any longer detect problematic spans/queries/patterns because every single query looks the same. 

All things considered, I think truncating the query to first word and using that as the SQL statement is a very good tradeoff IMO with very little downsides if any at all.

Reply:
Since we need to get these attributes from somewhere, unless there is another solution that does even less parsing/truncation/etc. we should use the solution implemented here. Nevertheless, I understand how this can look as going against the specification (which by the way does not clearly specify if not doing any parsing is a RECOMMEND NOT or SHOULD NOT type of requirement), so, @owais can you add a comment that explains your point of view here?

Reply:
Added